File: src/falcon_core/communications/voltage_states/device_voltage_states.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.device_voltage_states import DeviceVoltageStates as _CDeviceVoltageStates
from falcon_core.physics.device_structures.connection import Connection
from falcon_core.communications.voltage_states.device_voltage_state import DeviceVoltageState
from falcon_core.generic.list import List
from falcon_core.math.point import Point

class DeviceVoltageStates:
    """Python wrapper for DeviceVoltageStates."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def new_empty(cls, ) -> DeviceVoltageStates:
        return cls(_CDeviceVoltageStates.new_empty())

    @classmethod
    def new(cls, items: List) -> DeviceVoltageStates:
        return cls(_CDeviceVoltageStates.new(items._c))

    @classmethod
    def from_json(cls, json: str) -> DeviceVoltageStates:
        return cls(_CDeviceVoltageStates.from_json(json))

    def states(self, ) -> List:
        ret = self._c.states()
        if ret is None: return None
        return List(ret)

    def add_state(self, state: DeviceVoltageState) -> None:
        ret = self._c.add_state(state._c)
        return ret

    def find_state(self, connection: Connection) -> DeviceVoltageStates:
        ret = self._c.find_state(connection._c)
        return cls._from_capi(ret)

    def to_point(self, ) -> Point:
        ret = self._c.to_point()
        if ret is None: return None
        return Point._from_capi(ret)

    def intersection(self, other: DeviceVoltageStates) -> DeviceVoltageStates:
        ret = self._c.intersection(other._c)
        return cls._from_capi(ret)

    def push_back(self, value: DeviceVoltageState) -> None:
        ret = self._c.push_back(value._c)
        return ret

    def size(self, ) -> None:
        ret = self._c.size()
        return ret

    def empty(self, ) -> None:
        ret = self._c.empty()
        return ret

    def erase_at(self, idx: Any) -> None:
        ret = self._c.erase_at(idx)
        return ret

    def clear(self, ) -> None:
        ret = self._c.clear()
        return ret

    def const_at(self, idx: Any) -> DeviceVoltageState:
        ret = self._c.const_at(idx)
        if ret is None: return None
        return DeviceVoltageState._from_capi(ret)

    def at(self, idx: Any) -> DeviceVoltageState:
        ret = self._c.at(idx)
        if ret is None: return None
        return DeviceVoltageState._from_capi(ret)

    def items(self, ) -> List:
        ret = self._c.items()
        if ret is None: return None
        return List(ret)

    def contains(self, value: DeviceVoltageState) -> None:
        ret = self._c.contains(value._c)
        return ret

    def index(self, value: DeviceVoltageState) -> None:
        ret = self._c.index(value._c)
        return ret

    def equal(self, b: DeviceVoltageStates) -> None:
        ret = self._c.equal(b._c)
        return ret

    def not_equal(self, b: DeviceVoltageStates) -> None:
        ret = self._c.not_equal(b._c)
        return ret

    def __add__(self, other):
        """Operator overload for +"""
        if hasattr(other, "_c") and type(other).__name__ == "State":
            return self.add_state(other)
        return NotImplemented

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, DeviceVoltageStates):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, DeviceVoltageStates):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/src/communications/voltage_states/devicevoltagestates.ml
open Ctypes
open Capi_bindings
open Error_handling

(* No opens needed - using qualified names *)

class c_devicevoltagestates (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.devicevoltagestates_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module DeviceVoltageStates = struct
  type t = c_devicevoltagestates

  let empty () : t =
    let ptr = Capi_bindings.devicevoltagestates_create_empty () in
    ErrorHandling.raise_if_error ();
    new c_devicevoltagestates ptr

  let make (items : Listdevicevoltagestate.t) : t =
    ErrorHandling.read items (fun () ->
      let ptr = Capi_bindings.devicevoltagestates_create items#raw in
      ErrorHandling.raise_if_error ();
      new c_devicevoltagestates ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.devicevoltagestates_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_devicevoltagestates ptr

  let states (handle : t) : Listdevicevoltagestate.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.devicevoltagestates_states handle#raw in
      ErrorHandling.raise_if_error ();
      new c_listdevicevoltagestate result
    )

  let addState (handle : t) (state : Devicevoltagestate.t) : unit =
    ErrorHandling.multi_read [handle; state] (fun () ->
      let result = Capi_bindings.devicevoltagestates_add_state handle#raw state#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let findState (handle : t) (connection : Connection.t) : t =
    ErrorHandling.multi_read [handle; connection] (fun () ->
      let result = Capi_bindings.devicevoltagestates_find_state handle#raw connection#raw in
      ErrorHandling.raise_if_error ();
      new c_devicevoltagestates result
    )

  let toPoint (handle : t) : Point.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.devicevoltagestates_to_point handle#raw in
      ErrorHandling.raise_if_error ();
      new c_point result
    )

  let intersection (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.devicevoltagestates_intersection handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_devicevoltagestates result
    )

  let pushBack (handle : t) (value : Devicevoltagestate.t) : unit =
    ErrorHandling.multi_read [handle; value] (fun () ->
      let result = Capi_bindings.devicevoltagestates_push_back handle#raw value#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let size (handle : t) : int =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.devicevoltagestates_size handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let empty (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.devicevoltagestates_empty handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let eraseAt (handle : t) (idx : int) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.devicevoltagestates_erase_at handle#raw (Unsigned.Size_t.of_int idx) in
      ErrorHandling.raise_if_error ();
      result
    )

  let clear (handle : t) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.devicevoltagestates_clear handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let at (handle : t) (idx : int) : Devicevoltagestate.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.devicevoltagestates_at handle#raw (Unsigned.Size_t.of_int idx) in
      ErrorHandling.raise_if_error ();
      new c_devicevoltagestate result
    )

  let items (handle : t) : Listdevicevoltagestate.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.devicevoltagestates_items handle#raw in
      ErrorHandling.raise_if_error ();
      new c_listdevicevoltagestate result
    )

  let contains (handle : t) (value : Devicevoltagestate.t) : bool =
    ErrorHandling.multi_read [handle; value] (fun () ->
      let result = Capi_bindings.devicevoltagestates_contains handle#raw value#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let index (handle : t) (value : Devicevoltagestate.t) : int =
    ErrorHandling.multi_read [handle; value] (fun () ->
      let result = Capi_bindings.devicevoltagestates_index handle#raw value#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let equal (a : t) (b : t) : bool =
    ErrorHandling.multi_read [a; b] (fun () ->
      let result = Capi_bindings.devicevoltagestates_equal a#raw b#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (a : t) (b : t) : bool =
    ErrorHandling.multi_read [a; b] (fun () ->
      let result = Capi_bindings.devicevoltagestates_not_equal a#raw b#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.devicevoltagestates_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

end